File: java_version/Roslina.java
public abstract class Roslina extends Organizm {
    protected boolean czyRozprzesrzeniono = false;

    public Roslina(int sila, int x, int y, Swiat swiat) {
        super(sila, 0, x, y, swiat);
    }

    @Override
    public void akcja() {
        if ((int)(Math.random() * 5) == 4) {
            rozprzestrzenianie();
        }
    }

    @Override
    public void kolizja(Organizm inny) {
        swiat.dodajLog(getNazwa() + " na polu (" + x + ", " + y + ") został zjedzony przez " + inny.getNazwa());
        this.zabij();
    }

    protected void rozprzestrzenianie() {
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        for (int i = 0; i < 4; ++i) {
            int nowyX = x + dx[i];
            int nowyY = y + dy[i];

            if (swiat.czyWPlanszy(nowyX, nowyY) && swiat.czyPolePuste(nowyX, nowyY)) {
                zasiej(nowyX, nowyY);
                czyRozprzesrzeniono = true;
                return;
            }
        }
    }

    protected abstract void zasiej(int nowyX, int nowyY);

    public abstract String getNazwa();
    public abstract char rysowanie();
}

File: python_version/Roslina.py
from Organizm import Organizm
import random

class Roslina(Organizm):
    def __init__(self, sila, x, y, swiat):
        super().__init__(sila, 0, x, y, swiat)

    def akcja(self):
        if random.randint(0, 4) == 4:
            self.rozprzestrzenianie()

    def kolizja(self, inny):
        self.swiat.dodaj_log(f"{self.get_nazwa()} na ({self.x}, {self.y}) został zjedzony przez {inny.get_nazwa()}")
        self.zabij()

    def rozprzestrzenianie(self):
        for dx, dy in [(-1,0), (1,0), (0,-1), (0,1)]:
            nowy_x = self.x + dx
            nowy_y = self.y + dy
            if self.swiat.czy_pole_puste(nowy_x, nowy_y):
                self.zasiej(nowy_x, nowy_y)
                return

    def zasiej(self, nowy_x, nowy_y):
        raise NotImplementedError()
